com.unity.perception/Runtime/Randomization/Randomizers/AssetSources/AssetSource.cs

```diff
@@ -0,0 +1,139 @@
+﻿using System;
+using UnityEngine.Perception.Randomization.Samplers;
+
+namespace UnityEngine.Perception.Randomization
+{
+    /// <summary>
+    /// AssetSources are used to load assets from a generically within a <see cref="Randomizers.Randomizer"/>
+    /// </summary>
+    /// <typeparam name="T">The type of asset to load</typeparam>
+    [Serializable]
+    public sealed class AssetSource<T> where T : Object
+    {
+        [SerializeReference] ArchetypeBase m_ArchetypeBase;
+
+        /// <summary>
+        /// The location to load assets from
+        /// </summary>
+        [SerializeReference] public AssetSourceLocation assetSourceLocation = new LocalAssetSourceLocation();
+
+        bool m_Initialized;
+        UniformSampler m_Sampler = new UniformSampler();
+
+        /// <summary>
+        /// The archetype used to preprocess assets from this source
+        /// </summary>
+        public Archetype<T> archetype
+        {
+            get => (Archetype<T>)m_ArchetypeBase;
+            set => m_ArchetypeBase = value;
+        }
+
+        /// <summary>
+        /// The number of assets available within this asset source
+        /// </summary>
+        public int Count => assetSourceLocation.Count;
```

Comment:
properties should be camelCase (first letter lower)

Reply:
I agree, but I didn't want to break convention with the existing list class in C#. I've got no problem with switching it though.

Reply:
Counts have been changed to count